File: zig/src/bitmap_loader.zig
const std = @import("std");
const Color = @import("color.zig").Color;
const bitmap_mod = @import("bitmap.zig");
const Bitmap = bitmap_mod.Bitmap;
const Sprite = bitmap_mod.Sprite;
const Allocator = std.mem.Allocator;

pub const BitmapLoader = struct {
    allocator: Allocator,
    search_paths: []const []const u8,
    cache: std.StringHashMap(Bitmap),

    pub fn init(allocator: Allocator, search_paths: []const []const u8) BitmapLoader {
        return .{
            .allocator = allocator,
            .search_paths = search_paths,
            .cache = std.StringHashMap(Bitmap).init(allocator),
        };
    }

    pub fn deinit(self: *BitmapLoader) void {
        var it = self.cache.iterator();
        while (it.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
            self.allocator.free(entry.value_ptr.pixels);
        }
        self.cache.deinit();
    }

    pub fn loadBitmap(self: *BitmapLoader, name: []const u8) !Bitmap {
        if (self.cache.get(name)) |bmp| {
            return bmp;
        }

        const file_path = try self.findBitmapFile(name) orelse
            return error.BitmapNotFound;

        const bmp = try self.loadBitmapFromFile(file_path);
        self.allocator.free(file_path);
        const name_copy = try self.allocator.dupe(u8, name);
        try self.cache.put(name_copy, bmp);
        return bmp;
    }

    pub fn loadSprite(self: *BitmapLoader, name: []const u8) !Sprite {
        const bmp = try self.loadBitmap(name);
        return Sprite{ .bitmap = bmp };
    }

    fn findBitmapFile(self: *BitmapLoader, name: []const u8) !?[]u8 {
        for (self.search_paths) |dir| {
            const path = try std.fmt.allocPrint(self.allocator, "{s}/{s}.txt", .{ dir, name });
            std.fs.cwd().access(path, .{}) catch {
                self.allocator.free(path);
                continue;
            };
            return path;
        }
        return null;
    }

    fn loadBitmapFromFile(self: *BitmapLoader, path: []const u8) !Bitmap {
        const file = try std.fs.cwd().openFile(path, .{});
        defer file.close();

        const content = try file.readToEndAlloc(self.allocator, 1024 * 1024);
        defer self.allocator.free(content);

        // Parse lines to find dimensions
        var lines: std.ArrayList([]const u8) = .empty;
        defer lines.deinit(self.allocator);

        var width: u32 = 0;
        var line_start: usize = 0;
        for (content, 0..) |c, i| {
            if (c == '\n' or i == content.len - 1) {
                var line_end = i;
                if (c != '\n' and i == content.len - 1) {
                    line_end = i + 1;
                }
                var line = content[line_start..line_end];
                // Strip trailing '|' and '\r'
                while (line.len > 0 and (line[line.len - 1] == '|' or line[line.len - 1] == '\r')) {
                    line = line[0 .. line.len - 1];
                }
                if (line.len > width) {
                    width = @intCast(line.len);
                }
                try lines.append(self.allocator, line);
                line_start = i + 1;
            }
        }

        const height: u32 = @intCast(lines.items.len);
        const pixels = try self.allocator.alloc(Color, width * height);
        @memset(pixels, .transparent);

        for (lines.items, 0..) |line, y| {
            const max_x = @min(width, @as(u32, @intCast(line.len)));
            for (0..max_x) |x| {
                if (mapColor(line[x])) |color| {
                    pixels[y * width + x] = color;
                }
            }
        }

        return Bitmap{
            .width = width,
            .height = height,
            .pixels = pixels,
        };
    }
};

fn mapColor(c: u8) ?Color {
    return switch (c) {
        ' ' => .transparent,
        '.', '0' => .black,
        '#', '1' => .white,
        '2' => .red,
        '3' => .cyan,
        '4' => .purple,
        '5' => .green,
        '6' => .blue,
        '7' => .yellow,
        '8' => .orange,
        '9' => .brown,
        'a' => .light_red,
        'b' => .dark_grey,
        'c' => .grey,
        'd' => .light_green,
        'e' => .light_blue,
        'f' => .light_grey,
        else => null,
    };
}

File: ts/src/bitmap/bitmap-loader.ts
import { readFileSync, existsSync } from "node:fs";
import type { Bitmap } from "./bitmap.js";
import { SimpleBitmap } from "./simple-bitmap.js";
import { Sprite } from "./sprite.js";
import { Color } from "../server/color.js";

const colorMap: Record<string, Color> = {
  " ": Color.TRANSPARENT,
  ".": Color.BLACK,
  "0": Color.BLACK,
  "#": Color.WHITE,
  "1": Color.WHITE,
  "2": Color.RED,
  "3": Color.CYAN,
  "4": Color.PURPLE,
  "5": Color.GREEN,
  "6": Color.BLUE,
  "7": Color.YELLOW,
  "8": Color.ORANGE,
  "9": Color.BROWN,
  a: Color.LIGHT_RED,
  b: Color.DARK_GREY,
  c: Color.GREY,
  d: Color.LIGHT_GREEN,
  e: Color.LIGHT_BLUE,
  f: Color.LIGHT_GREY,
};

export { colorMap };

export class BitmapLoader {
  private bitmapCache = new Map<string, Bitmap>();
  private readonly bitmapPath: string[];

  constructor(bitmapPath: string) {
    this.bitmapPath = bitmapPath.split(":");
  }

  loadSprite(name: string, x = 0, y = 0): Sprite {
    const bitmap = this.loadBitmap(name);
    return new Sprite(bitmap, x, y);
  }

  loadBitmap(name: string): Bitmap {
    if (!this.bitmapCache.has(name)) {
      const file = this.findBitmapFileInPath(name);
      if (file) {
        this.bitmapCache.set(name, this.loadBitmapFromFile(file));
      }
    }
    const bitmap = this.bitmapCache.get(name);
    if (!bitmap) {
      throw new Error(`bitmap not found: ${name}`);
    }
    return bitmap;
  }

  private loadBitmapFromFile(file: string): Bitmap {
    const content = readFileSync(file, "utf-8");
    const lines: string[] = [];
    let width = 0;
    for (const rawLine of content.split("\n")) {
      const line = rawLine.replace(/[|\r\n]+$/, "");
      if (line.length > width) {
        width = line.length;
      }
      lines.push(line);
    }
    // Remove trailing empty line if present
    while (lines.length > 0 && lines[lines.length - 1] === "") {
      lines.pop();
    }
    const height = lines.length;
    const pixels = new Array(width * height).fill(Color.TRANSPARENT);
    for (let y = 0; y < height; y++) {
      const line = lines[y];
      const maxX = Math.min(width, line.length);
      for (let x = 0; x < maxX; x++) {
        const ch = line[x];
        if (ch in colorMap) {
          pixels[y * width + x] = colorMap[ch];
        }
      }
    }
    return new SimpleBitmap(width, height, pixels);
  }

  private findBitmapFileInPath(bitmapName: string): string | null {
    for (const dir of this.bitmapPath) {
      const txtFile = `${dir}/${bitmapName}.txt`;
      if (existsSync(txtFile)) {
        return txtFile;
      }
    }
    return null;
  }
}
